File: week7/java/src/seonghoo1217/_2512.java
package seonghoo1217;

import java.io.*;
import java.util.StringTokenizer;

public class _2512 {
	//19m01s48

	public static void main(String[] args) throws IOException {
	    BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
	    BufferedWriter bw=new BufferedWriter(new OutputStreamWriter(System.out));
		int country = Integer.parseInt(br.readLine());
		int [] costs=new int[country];

		long left=0,right=-1;
		StringTokenizer st=new StringTokenizer(br.readLine());
		for (int i=0;i<country;i++){
			costs[i]=Integer.parseInt(st.nextToken());
			right=Math.max(right,costs[i]);
		}

		int budget = Integer.parseInt(br.readLine());
		long answer=0;
		while (left<=right){
			long mid=(left+right)/2;
			long sum=0;
			for (int money:costs){
				if (money>=mid) sum+=mid;
				else sum+=money;
			}

			if (sum>budget){
				right=mid-1;
			}else {
				left=mid+1;
				answer=Math.max(answer,mid);
			}
		}
		System.out.println(right);
	}
}

File: week7/javascript/Siby1lA/_2512.js
//00:17:14
function solution(case1) {
  const input = case1.split("\n");
  const [N, R, M] = [+input[0], input[1].split(" ").map(Number), +input[2]];
  R.sort((a, b) => a - b);
  let left = 0;
  let right = R[N - 1];
  let answer = Number.MIN_SAFE_INTEGER;
  while (left <= right) {
    let mid = Math.floor((left + right) / 2);

    let sum = 0;
    for (let x of R) {
      if (x > mid) sum += mid;
      else sum += x;
    }
    if (M >= sum) {
      if (mid > answer) answer = mid;
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return answer;
}
const case1 = `4
120 110 140 150
485`;

console.log(solution(case1));
